src/webgpu/shader/validation/expression/call/builtin/transpose.spec.ts

```diff
@@ -0,0 +1,90 @@
+const builtin = 'transpose';
+export const description = `
+Validation tests for the ${builtin}() builtin.
+`;
+
+import { makeTestGroup } from '../../../../../../common/framework/test_group.js';
+import { keysOf, objectsToRecord } from '../../../../../../common/util/data_tables.js';
+import { Type, kAllMatrices, scalarTypeOf } from '../../../../../util/conversion.js';
+import { ShaderValidationTest } from '../../../shader_validation_test.js';
+
+import {
+  fullRangeForType,
+  kConstantAndOverrideStages,
+  stageSupportsType,
+  validateConstOrOverrideBuiltinEval,
+} from './const_override_validation.js';
+
+export const g = makeTestGroup(ShaderValidationTest);
+
+const kValidArgumentTypes = objectsToRecord(kAllMatrices);
+
+g.test('values')
+  .desc(
+    `
+Validates that constant evaluation and override evaluation of ${builtin}() error on invalid inputs.
+`
+  )
+  .params(u =>
+    u
+      .combine('stage', kConstantAndOverrideStages)
+      .combine('type', keysOf(kValidArgumentTypes))
+      .filter(u => stageSupportsType(u.stage, kValidArgumentTypes[u.type]))
+      .beginSubcases()
+      .expand('value', u => fullRangeForType(kValidArgumentTypes[u.type]))
+  )
+  .beforeAllSubcases(t => {
+    if (scalarTypeOf(kValidArgumentTypes[t.params.type]) === Type.f16) {
+      t.selectDeviceOrSkipTestCase('shader-f16');
+    }
+  })
+  .fn(t => {
+    const expectedResult = true;
+
+    const type = kValidArgumentTypes[t.params.type];
+    validateConstOrOverrideBuiltinEval(
+      t,
+      builtin,
+      expectedResult,
+      [type.create(t.params.value)],
+      t.params.stage
+    );
+  });
+
+const kArgCases = {
+  good: '(mat2x2(0, 1, 2, 3))',
+  bad_no_parens: '',
+  // Bad number of args
+  bad_0args: '()',
+  bad_2arg: '(1.2, 2.3)',
+  // Bad value for arg 0
+  bad_0bool: '(false)',
+  bad_0array: '(array(1.1,2.2))',
+  bad_0struct: '(modf(2.2))',
+  bad_0uint: '(1u)',
+  bad_0int: '(1i)',
+  bad_0vec2i: '(vec2i())',
+  bad_0vec2u: '(vec2u())',
+  bad_0vec3i: '(vec3i())',
+  bad_0vec3u: '(vec3u())',
+  bad_0vec4i: '(vec4i())',
+  bad_0vec4u: '(vec4u())',
+};
+
+g.test('args')
+  .desc(`Test compilation failure of ${builtin} with variously shaped and typed arguments`)
+  .params(u => u.combine('arg', keysOf(kArgCases)))
+  .fn(t => {
+    t.expectCompileResult(
+      t.params.arg === 'good',
+      `const c = ${builtin}${kArgCases[t.params.arg]};`
+    );
+  });
+
```

Comment:
On https://github.com/gpuweb/cts/pull/3626 David pointed me at https://github.com/gpuweb/cts/pull/3612/files#diff-fe62cb0fc68ba76c7b93bf333a9a9542cdb242f14292bab49db741f6b4f72409R46 as an example of another type of validation test that we should be adding for the return types.

Seems especially relevant here, given the relationship between the input type and return type.

Reply:
I agree.
I feel this test should iterate over all matrix types for inputs an for outputs, and the transpose return type should be checked.

```
for input-scalar in abstract-float, f32, f16:
   for input-C in 2,3,4 
     for input-R in 2,3,4
        for output-scalar in  f32, f16:
           for output-c in 2,3,4
               for output-r in 2,3,4
                 check   something like `const c : output-mat-type = transpose(input-mat-type)` compiles or doesn't compile based on convertibility.  input-scalar-to-output-scalar, and output-r = input-C and output-c = input-r.
```

Note you can use isConvertible from src/webpgu/util/conversion.ts   (I'd probably only check convertibility of the scalar types.)

Reply:
Done